sessions/2018/july/Session_9.md

```diff
@@ -0,0 +1,124 @@
+# SOFTWARE ENGINEERING IN JAVA
+
+## Session 9 (07/08/2018)
+
+- TDD
+
+### Notes:
+
+#### TDD
+
+There exists 3 types of Test framework (terms can vary): *Unit*, *component* and *integration*.
+
+##### Unit Tests
+Unit tests are used for only test a single class.
+
+##### Component Tests
+Component tests are very fast and are intended for multiple clases but not for I/O or databases. A common pattern of this is to test a package.
+
+##### Integration Tests
+Integration tests are used to test classes and databases. In this category we can found two tendencies:
+- System tests: A separate project test our project.
+- Acceptance tests: A guided test for manual tests, can be automated, but in general terms is slow.
+
+TDD is a discipline that we have to stick to, following these laws we will end with good code.
+
+In TDD we implement the simplest solution we can.
+
+The results using TDD shoud be more elegant and better than any clever idea (we can write our idea and after compare the results).
+
+We program in order to provoke crashes, so we can fix them before they occurs in the reality. There are 3 steps to follow:
+
+- Red
+- Green
+- Refactor
+
+##### <span style="color:red;">Red</span>
+Write as little code as posible for make a failing test. Has to be relevant to the test that we're doing. A compilation error is a valid case.
```

Comment:
Write as little *TEST* code as possible. You're not allowed to write production code here.